### PR TITLE
Properly close websockets on server shutdown.

### LINKED ISSUE #222: Netgraph takes super long in Firefox to load
no idea why ...

### PR DESCRIPTION
**Pull request by jgosmann**
Fixes #222.

I think, with this PR all websocket connections are always properly closed no matter which side initiates the close. Because of the organization of the server code, this patch is not as straight-forward as I'd like it to be. In fact, the organization of the server code seems a bit convoluted without a good separation of concerns. But refactoring it would be a slightly larger job. It might be complicated by the base Python classes which probably where to designed with prolonged connections (websockets) in mind.

### CODE REVIEW

File: nengo_viz/swi.py [hunk 1]
```diff
@@ -520,6 +535,9 @@ def __init__(self, socket, addr):
         self.socket = socket
         self.addr = addr
         self.buffered_data = bytearray([])
+        self.remote_close = False;
+        self._closing = False;
+        self._closed = False;
```
Comment (hunse) [inline nengo_viz/swi.py]: You've; been; writing; too; much; javascript;

File: nengo_viz/swi.py [hunk 2]
```diff
@@ -565,9 +583,15 @@ def read(self):
         datalen = data[1] & 0x7F
 
         if opcode == 8:
-            code = 0b10001000
-            ack = struct.pack('!BB', code, 0)
-            self.socket.send(ack)
+            if self._closing:
+                self._closed = True;
```
Comment (hunse) [inline nengo_viz/swi.py]: Bring this outside the `if` block since it gets set to true either way.

File: nengo_viz/swi.py [hunk 3]
```diff
@@ -211,7 +213,7 @@ def handle_ws_request(self, args, db):
         else:
             command = 'ws_%s' % args[0]
 
-        client = ClientSocket(self.request, '')
+        client = self.server.create_websocket(self.request)
```
Comment (hunse) [inline nengo_viz/swi.py]: Where does `self.server` come from?

File: nengo_viz/swi.py [hunk 4]
```diff
@@ -565,9 +583,15 @@ def read(self):
...
-            code = 0b10001000
-            ack = struct.pack('!BB', code, 0)
-            self.socket.send(ack)
+            if self._closing:
+                self._closed = True;
+            else:
+                self.remote_close = True;
+                self._closing = True;
+                self._closed = True;
+                self.close()  # acknowledge the close request
```
Comment (hunse) [inline nengo_viz/swi.py]: This will set `self._closing`, so no need to set it here, right?

File: nengo_viz/swi.py (hunk 1 above)
Comment (jgosmann) [inline nengo_viz/swi.py]: yes;

File: nengo_viz/swi.py (hunk 3 above)
Comment (jgosmann) [inline nengo_viz/swi.py]: `BaseHTTPRequestHandler`

**Comment by jgosmann** [comment]
Fixed that stuff.

**Comment by tbekolay** [comment]
The code looks good... but I get a bunch of errors when I close all the open tabs. Here's the traceback, if it helps:

```bash
<class 'socket.error'> [Errno 107] Transport endpoint is not connected ['__class__', '__delattr__', '__dict__', '__doc__', '__format__', '__getattribute__', '__getitem__', '__getslice__', '__hash__', '__init__', '__module__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '__unicode__', '__weakref__', 'args', 'errno', 'filename', 'message', 'strerror']
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 40198)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 599, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 655, in __init__
    self.handle()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
    method()
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 201, in do_GET
    self.handle_request(args, db)
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 243, in handle_request
    self.handle_ws_request(args, db)
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 237, in handle_ws_request
    client.wait_for_close()
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 674, in wait_for_close
<class 'socket.error'> [Errno 107] Transport endpoint is not connected ['__class__', '__delattr__', '__dict__', '__doc__', '__format__', '__getattribute__', '__getitem__', '__getslice__', '__hash__', '__init__', '__module__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '__unicode__', '__weakref__', 'args', 'errno', 'filename', 'message', 'strerror']
    self.read()
----------------------------------------
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 590, in read
    self.socket.shutdown(socket.SHUT_RDWR)
  File "/usr/lib/python2.7/socket.py", line 228, in meth
Exception happened during processing of request from ('127.0.0.1', 40202)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 599, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 655, in __init__
    return getattr(self._sock,name)(*args)
    self.handle()
error: [Errno 107] Transport endpoint is not connected
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 340, in handle
----------------------------------------
    self.handle_one_request()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
    method()
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 201, in do_GET
    self.handle_request(args, db)
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 243, in handle_request
    self.handle_ws_request(args, db)
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 237, in handle_ws_request
    client.wait_for_close()
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 674, in wait_for_close
    self.read()
  File "/home/tbekolay/Code/nengo_viz/nengo_viz/swi.py", line 590, in read
    self.socket.shutdown(socket.SHUT_RDWR)
  File "/usr/lib/python2.7/socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
error: [Errno 107] Transport endpoint is not connected
----------------------------------------
No connections remaining to the nengo_viz server.
Shutting down server...
1 zombie threads will close abruptly
Unhandled exception in thread started by <bound method Thread.__bootstrap of <Thread(Thread-48, stopped daemon 139777965799168)>>Unhandled exception in thread started by <bound method Thread.__bootstrap of <Thread(Thread-3, stopped daemon 139778959537920)>>
Traceback (most recent call last):
```

**Comment by jgosmann** [comment]
What OS? It works for me on Mac.

**Comment by tbekolay** [comment]
Debian. FYI my sequence of steps was to open a tab in Chrome, open a tab in FF, play and pause the sim in Chrome, play and pause in FF, close the FF tab, then close the Chrome tab.

**Comment by jgosmann** [comment]
Should be fixed.

**Comment by tbekolay** [comment]
Cool, that fixed it for me. No more tracebacks!

**Comment by jgosmann** [comment]
It might not be too bad. We should try polling with `select` whether the socket is ready for writing and if not just drop the message.

**Comment by tbekolay** [comment]
Are there any outstanding issues on this one, or is it this ready for rebase / merge?

**Comment by jgosmann** [comment]
Ah, here's the comment I was missing m(
Posted into the wrong thread

**Comment by tbekolay** [comment]
Classic Jan

**Comment by jgosmann** [comment]
Rebasing this now

**Comment by jgosmann** [comment]
Rebased

**Pull request merged**